core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala

```diff
@@ -267,12 +277,17 @@ private[spark] class EventLoggingListener(
 
   override def onExecutorMetricsUpdate(event: SparkListenerExecutorMetricsUpdate): Unit = {
     if (shouldLogStageExecutorMetrics) {
-      // For the active stages, record any new peak values for the memory metrics for the executor
-      event.executorUpdates.foreach { executorUpdates =>
-        liveStageExecutorMetrics.values.foreach { peakExecutorMetrics =>
-          val peakMetrics = peakExecutorMetrics.getOrElseUpdate(
-            event.execId, new ExecutorMetrics())
-          peakMetrics.compareAndUpdatePeakValues(executorUpdates)
+      event.executorUpdates.foreach { case (stageKey1, peaks) =>
+        liveStageExecutorMetrics.foreach { case (stageKey2, metricsPerExecutor) =>
+          // If the update came from the driver, stageKey1 will be the dummy key (-1, -1),
+          // so record those peaks for all active stages.
+          // Otherwise, record the peaks for the matching stage.
+          val driverStageKey = (-1, -1)
+          if (stageKey1 == driverStageKey || stageKey1 == stageKey2) {
+            val metrics = metricsPerExecutor.getOrElseUpdate(
+              event.execId, new ExecutorMetrics())
+            metrics.compareAndUpdatePeakValues(peaks)
```

Comment:
ok I think I understand what is happening here now, sorry for my confusion earlier.

I'm a bit concerned about this approach though ... events can get backlogged quite a bit on the listenerbus, so the stages running now might not be the stages running when the driver collected the metrics.  I don't have a suggestion for what we could do that would be better at the moment, though, I'll need to think about it more.

Reply:
so the trick here is having the DAGScheduler expose the set of running stages in a thread-safe manner.  a ConcurrentSkipListSet might be a good choice, but I will keep thinking about this.

Reply:
ok, nevermind all of the above -- I take back my statement about the running stages being out of sync.  It is true that the dag scheduler may be running some other stages at this point, but that doesn't matter -- this is using the running stages as viewed by the EventLoggingListener.  there may be a small discrepancy with the order in which the metric poller pushes the metrics vs the dagscheduler pushes the stage update, but we'd have to live with that no matter what.